File: Cognito/decode-verify-jwt/Node.js/decode-verify-jwt.js
var https = require('https');
var jose = require('node-jose');

var region = 'ap-southeast-2';
var userpool_id = 'ap-southeast-2_xxxxxxxxx';
var keys_url = 'https://cognito-idp.' + region + '.amazonaws.com/' + userpool_id + '/.well-known/jwks.json';

exports.handler = (event, context, callback) => {
    var token = event['token'];
    var sections = token.split('.');
    // get the kid from the headers prior to verification
    var header = jose.util.base64url.decode(sections[0]);
    header = JSON.parse(header);
    var kid = header['kid'];
    // download the public keys
    https.get(keys_url, function(response) {
        if (response.statusCode == 200) {
            response.on('data', function(body) {
                var keys = JSON.parse(body)['keys'];
                // search for the kid in the downloaded public keys
                var key_index = -1;
                for (var i=0; i < keys.length; i++) {
                        if (kid == keys[i].kid) {
                            key_index = i;
                            break;
                        }
                }
                if (key_index == -1) {
                    callback('Public key not found in jwks.json');
                }
                // construct the public key
                jose.JWK.asKey(keys[key_index]).
                then(function(result) {
                    // verify the signature
                    jose.JWS.createVerify(result).
                    verify(token).
                    then(function(result) {
                        // do some stuff with our claims
                        callback(null, JSON.parse(result.payload));
                    }).
                    catch(function() {
                        callback('Signature verification failed');
                    });
                });
            });
        }
    });
}

File: Cognito/decode-verify-jwt/decode-verify-jwt.go
package main

import (
	"context"
	"errors"
	"fmt"
	"log"
	"os"

	"github.com/golang-jwt/jwt/v4"
	"github.com/lestrrat/go-jwx/jwk"
)

var region string
var userpool_id string
var app_client_id string
var keys_url string

var keys *jwk.Set

type AWSCognitoClaims struct {
	ClientId string `json:client_id`
	Username string `json:username`
	jwt.StandardClaims
}

func init() {

	region = os.Getenv("AWS_REGION")
	userpool_id = os.Getenv("USERPOOL_ID")
	app_client_id = os.Getenv("APP_CLIENT_ID")
	keys_url = fmt.Sprintf("https://cognito-idp.%v.amazonaws.com/%v/.well-known/jwks.json", region, userpool_id)

	var err error
	keys, err = jwk.FetchHTTP(keys_url)
	if err != nil {
		log.Fatal(err)
	}
}

func HandleRequest(ctx context.Context, event map[string]string) (jwt.Claims, error) {

	token_string := event["token"]
	claims := &AWSCognitoClaims{}

	// Parse and validate received token
	_, err := jwt.ParseWithClaims(token_string, claims, validateToken)
	if err != nil {
		return nil, err
	}

	return claims, nil
}

func validateToken(token *jwt.Token) (interface{}, error) {

	// get the kid from the headers prior to verification
	keyID, ok := token.Header["kid"].(string)
	if !ok {
		return nil, errors.New("expecting JWT header to have string kid")
	}

	//  search for the kid in the downloaded public keys
	if key := keys.LookupKeyID(keyID); len(key) == 1 {

		claims := token.Claims.(*AWSCognitoClaims)
		// verify the Audience (use claims.ClientId if verifying an access token)
		if claims.Audience != app_client_id {
			return nil, errors.New("token was not issued for this audience")
		}

		// construct the public key
		return key[0].Materialize()

	}

	return nil, fmt.Errorf("unable to find key %q", keyID)
}

func main() {
	event := make(map[string]string)

	// for testing locally you can enter the JWT ID Token here
	event["token"] = ""

	HandleRequest(context.Background(), event)
	// lambda.Start(HandleRequest)
}
